### ISSUE TITLE
Custom type (struct/enum) smuggled into IR illegally via tuples.

### ISSUE THREAD

**Issue by otrho** [FuelLabs/sway#1556]
This is from @Braqzen in #1401 where it's not really related to that issue, so I'm recreating it here.

The following
```rust
contract;

abi MyContract {
    fn bug1(param: (Person, u64)) -> bool;
    fn bug2(param: (Location, u64)) -> bool;
}

struct Person {
    age: u64
}

enum Location {
    earth: ()
}

impl MyContract for Contract {
    fn bug1(param: (Person, u64)) -> bool {
        true
    }

    fn bug2(param: (Location, u64)) -> bool {
        true
    }
}
```

generates the error:
![1](https://user-images.githubusercontent.com/103777923/168435729-4ac0d2b1-c402-499e-82f1-19896034cb95.png)

The issue is with the tuple taking a `struct` / `enum` type. The `u64` works as expected. Using other `unsigned integers`, `str[<size>]`, `bool` and `b256` types works as expected when passed into the tuple.

**Comment by mohammadfawaz** [comment]
+ 1 for the title 😆

**Comment by emilyaherbert** [comment]
I believe that the type engine and monomorphization improvements from #579 will completely (or at least partially), fix this

**Comment by emilyaherbert** [comment]
Just confirmed---it _is_ closed by #579

**Comment by otrho** [comment]
Thanks -- when you mentioned it in the sync call I had my hopes up. :slightly_smiling_face:
